File: source/haskell-examples/Esercizi.hs
{- Esercizio 1 -}

-- trova il massimo di una coppia di numeri
massimo :: (Num a, Ord a) => (a,a) -> a
massimo (x, y)
    | x >= y = x
    | otherwise = y

-- data una lista di coppie di numeri ritorna una lista dei massimi di ogni coppia
listaMassimi :: (Num a, Ord a) => [(a,a)] -> [a]
listaMassimi = map massimo 

-- alternativamente possiamo usare la funzione built-in maximum
listaMassimi' :: (Num a, Ord a) => [(a,a)] -> [a]
listaMassimi' = map maximum

-- senza usare la funzione map
listaMassimiRic :: (Num a, Ord a) => [(a,a)] -> [a]
listaMassimiRic [] = []
listaMassimiRic (x:xs) = massimo x : listaMassimiRic xs

-- solo ricorsione
listaMassimiRic' :: (Num a, Ord a) => [(a,a)] -> [a]
listaMassimiRic' [] = []
listaMassimiRic' ((x,y):pairs) 
    | x >= y = x : listaMassimiRic' pairs
    | otherwise = y : listaMassimiRic' pairs

{- Esercizio 2 -}

-- Media di una lista di numeri
-- NB: la typeclass fractional serve per divisioni su numeri reali
-- NB: fromIntegral casta il risultato di lenght (classe Integral) alla classe Num
average :: (Fractional a) => [a] -> a
average xs = sum xs / fromIntegral (length xs)

{- Esercizio 3 -}

-- dividi per 400, se resto 0 è bisestile
-- altrimenti dividi per 4 e dividi per 100, se entrambe 0 è bisestile

nonBisestile :: Int -> Bool
nonBisestile x 
    | mod x 400 == 0 = False
    | (mod x 4 == 0) && not(mod x 100 == 0) = False 
    | otherwise = True

listaNonBisestili :: [Int] -> [Int]
listaNonBisestili = filter nonBisestile

File: source/prolog-examples/Esercizi.pl
/* 1) Lista dei massimi di ogni coppia */

% massimo di una coppia
coppia_massimo([X,Y], Max) :-
    number(X),
    number(Y),
    ((X >= Y, Max is X);
    (X < Y, Max is Y)).

% lista dei massimi di ogni coppia
lista_massimi([], []).
lista_massimi([H|T], [Nh|Nt]) :- 
    coppia_massimo(H, Nh),
    lista_massimi(T, Nt).

/* 2) Media di una lista di numeri */

% somma degli elementi
somma_lista([], 0).
somma_lista([H|T], S) :-
    number(H),
    somma_lista(T, Tmp),
    S is H + Tmp.

% media degli elementi
lista_media([], 0).
lista_media(L, M) :-
    somma_lista(L, S),
    length(L, Len),
    M is S / Len.

/* 3) Lista anni senza bisestili */

% stailire se un anno è bisestile
bisestile(Anno) :-
    T is Anno mod 400,
    T == 0.
bisestile(Anno) :-
    T1 is Anno mod 4,
    T2 is Anno mod 100,
    T1 == 0,
    T2 =\= 0.

% eliminare gli anni bisestili dalla lista
lista_non_bisestili([], []).
lista_non_bisestili([H|T], L) :-
    bisestile(H),
    lista_non_bisestili(T, L).
lista_non_bisestili([H|T], L) :-
    \+(bisestile(H)),
    lista_non_bisestili(T, L1),
    append(L1, [H], L).